File: 0075-sort-colors/0075-sort-colors.cpp
class Solution {
public:
    void sortColors(vector<int>& nums) {
        int len = nums.size();
        if (len == 0)
            return;

        int low = 0, high = len - 1;
        int temp;

        while (low != high) {
            while ((nums[low] == 0) && (low < high))
                low++;
            while ((nums[high] != 0) && (low < high))
                high--;
            if (low != high) {
                temp = nums[low];
                nums[low] = nums[high];
                nums[high] = temp;
            }
        }
        high = len - 1;
        while (low != high) {
            while ((nums[low] == 1) && (low < high))
                low++;
            while ((nums[high] != 1) && (low < high))
                high--;
            if (low != high) {
                temp = nums[low];
                nums[low] = nums[high];
                nums[high] = temp;
            }
        }
    }
};

File: 0075-sort-colors/0075-sort-colors.java
class Solution {
    void swap(int[] nums, int n1, int n2) {
        int temp = nums[n1];
        nums[n1] = nums[n2];
        nums[n2] = temp;
    }

    public void sortColors(int[] nums) {
        int n = nums.length;
        int zeroIndex = 0, firstIndex = 0, twoIndex = n - 1;

        while (firstIndex <= twoIndex) {
            if (nums[firstIndex] == 0) {
                swap(nums, zeroIndex, firstIndex);
                zeroIndex++;
                firstIndex++;
            } else if (nums[firstIndex] == 1) {
                firstIndex++;
            } else {
                swap(nums, firstIndex, twoIndex);
                twoIndex--;
            }
        }
    }
}